exchanges/stream/websocket.go

```diff
@@ -786,163 +776,134 @@ func (w *Websocket) GetName() string {
+func (w *Websocket) AddSuccessfulSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil {
+		return common.ErrNilPointer
+	}
 	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
+		w.subscriptions = subscription.NewStore()
 	}
-	for _, cN := range channels { //nolint:gocritic // See below comment
-		c := cN // cN is an iteration var; Not safe to make a pointer to
-		key := c.EnsureKeyed()
-		c.State = subscription.SubscribedState
-		w.subscriptions[key] = &c
+	var errs error
+	for _, s := range subs {
+		if err := s.SetState(subscription.SubscribedState); err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+		}
+		if err := w.subscriptions.Add(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
+	return errs
 }
 
-// RemoveSubscriptions removes subscriptions from the subscription list
-func (w *Websocket) RemoveSubscriptions(channels ...subscription.Subscription) {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
-	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
+// RemoveSubscriptions removes subscriptions from the subscription list and sets the status to Unsubscribed
+func (w *Websocket) RemoveSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil || w.subscriptions == nil {
+		return common.ErrNilPointer
 	}
-	for i := range channels {
-		key := channels[i].EnsureKeyed()
-		delete(w.subscriptions, key)
+	var errs error
+	for _, s := range subs {
+		if err := s.SetState(subscription.UnsubscribedState); err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+		}
+		if err := w.subscriptions.Remove(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
+	return errs
 }
 
-// GetSubscription returns a pointer to a copy of the subscription at the key provided
+// GetSubscription returns a subscription at the key provided
 // returns nil if no subscription is at that key or the key is nil
+// Keys can implement subscription.MatchableKey in order to provide custom matching logic
 func (w *Websocket) GetSubscription(key any) *subscription.Subscription {
-	if key == nil || w == nil || w.subscriptions == nil {
+	if w == nil || w.subscriptions == nil || key == nil {
 		return nil
 	}
-	w.subscriptionMutex.RLock()
-	defer w.subscriptionMutex.RUnlock()
-	if s, ok := w.subscriptions[key]; ok {
-		c := *s
-		return &c
-	}
-	return nil
+	return w.subscriptions.Get(key)
 }
 
 // GetSubscriptions returns a new slice of the subscriptions
-func (w *Websocket) GetSubscriptions() []subscription.Subscription {
-	w.subscriptionMutex.RLock()
-	defer w.subscriptionMutex.RUnlock()
-	subs := make([]subscription.Subscription, 0, len(w.subscriptions))
-	for _, c := range w.subscriptions {
-		subs = append(subs, *c)
+func (w *Websocket) GetSubscriptions() subscription.List {
+	if w == nil || w.subscriptions == nil {
+		return nil
```

Comment:
suggestion: return error

Reply:
Same rationale as before regarding errors on nils.
I'm swayed more to protect against them before, but not to the extent of adding extra errors just for them when something else *will* have caught the nil, and if it doesn't, this isn't going to introduce a panic.

But saying exactly that, it does mean we should return something iterable. So:

Fixed 44179b4